packages/back-end/src/services/informationSchema.ts

```diff
@@ -0,0 +1,270 @@
+            correspondingIndex
+          ].schemas[correspondingSchemaIndex].tables.findIndex(
+            (updatedTableRecord) =>
+              updatedTableRecord.tableName === table.tableName
+          );
+
+          if (correspondingTableIndex === -1) {
+            // This means that the table has been deleted.
+            deletedTableIds.push(table.id);
+          }
+        }
+      });
+    });
+  });
+  return deletedTableIds;
+}
+
+export async function mergeStaleInformationSchemaWithUpdate(
+  staleInformationSchema: InformationSchema[],
+  updatedInformationSchema: InformationSchema[],
+  organization: string
+): Promise<InformationSchema[]> {
+  // If there is no stale information schema, then return the updated information schema
+  // This could happen if there was an error when initially creating the informationSchema
+  if (!staleInformationSchema || staleInformationSchema.length === 0) {
+    return updatedInformationSchema;
+  }
+  updatedInformationSchema.forEach((database) => {
+    const correspondingIndex = staleInformationSchema.findIndex(
+      (staleInformationSchemaRecord) =>
+        staleInformationSchemaRecord.databaseName === database.databaseName
+    );
+    // If the database exists in the staleInformationSchemaArray, then update the dateUpdated
+    if (correspondingIndex > -1) {
+      database.dateCreated =
+        staleInformationSchema[correspondingIndex].dateCreated;
+    }
+    database.schemas.forEach((schema) => {
+      const correspondingSchemaIndex = staleInformationSchema[
+        correspondingIndex
+      ].schemas.findIndex(
+        (staleSchemaRecord) =>
+          staleSchemaRecord.schemaName === schema.schemaName
+      );
+
+      if (correspondingSchemaIndex > -1) {
+        schema.dateCreated =
+          staleInformationSchema[correspondingIndex].schemas[
+            correspondingSchemaIndex
+          ].dateCreated;
+      }
+      schema.tables.forEach(async (table) => {
+        const correspondingTableIndex = staleInformationSchema[
+          correspondingIndex
+        ].schemas[correspondingSchemaIndex].tables.findIndex(
+          (staleTableRecord) => staleTableRecord.tableName === table.tableName
+        );
+
+        if (correspondingTableIndex > -1) {
+          table.dateCreated =
+            staleInformationSchema[correspondingIndex].schemas[
+              correspondingSchemaIndex
+            ].tables[correspondingTableIndex].dateCreated;
+          table.id =
+            staleInformationSchema[correspondingIndex].schemas[
+              correspondingSchemaIndex
+            ].tables[correspondingTableIndex].id;
+          if (
+            table.numOfColumns ===
+            staleInformationSchema[correspondingIndex].schemas[
+              correspondingSchemaIndex
+            ].tables[correspondingTableIndex].numOfColumns
+          ) {
+            // If the number of columns hasn't changed, then we shoul set the dateUpdated to the stale date.
```

Comment:
This logic doesn't seem right.  Even if nothing changed, we still want to update the date.  Otherwise, aren't we going to keep trying to refresh it because we think its old, even though we just checked and nothing changed?

Reply:
This is updating the `informationSchema` itself, not the `informationSchemaTable` record. I built this logic so in the event an `informationSchema` as a whole is out of date, we're fetching a new version, and then updating the existing anywhere something is different.

And if nothing is different within any of the nested properties, we're not updating the nested `dateUpdated`. We are however, updating the `informationSchema`'s top-level `dateUpdated`.

This works because the `informationSchemaTable` record also has a top-level `dateUpdated` property, and that is used to refresh the actual table's metadata.

Reply:
To elaborate with an example:

Let's say we have an informationSchema that was last updated 32 days ago, when the user lands on the page, we're going to fetch a fresh version and compare it to the existing. In the event none of the schemas or tables were changed, the only date we'll update is the `informationSchema`'s top-level `dateUpdated`, but we'll leave all of the nested items as is, since they haven't been updated.

Now, let's say the org comes back in another 32 days, but this time, they've removed a column from a particular table. Now, we'll update the `informationSchema'`s top-level `dateUpdated` AND we'll update the nested table's `dateUpdated`. (We'll also update the `dateUpdated` property on the schema that contains the table that changed.

And all of this doesn't affect the actual table record that lives in the `InformationSchemaTables` collection as that record has it's own `dateCreated` and `dateUpdated`, and when a user clicks on that table to see the columns and metadata, we run a separate check there to see if that particular table is stale.

TBH, I'm not entirely sure that level of detail is required - we might be able to get away with just the top-level having a `dateCreated` and `dateUpdated`.